src/lib/customPropTypes.js

```diff
@@ -0,0 +1,182 @@
+      if (!_.isFunction(validator)) {
+        throw new Error(
+          `all() argument "validators" should contain functions, found: ${type(validator)}.`
+        )
+      }
+      return validator(props, propName, componentName, ...rest)
+    }))
+
+    // we can only return one error at a time
+    return errors[0]
+  }
+}
+
+/**
+ * Ensure a prop adherers to at least one of the given prop type validators.
+ * @param {function[]} validators An array of propType functions.
+ */
+export const some = (validators) => {
+  return (props, propName, componentName, ...rest) => {
+    if (!_.isArray(validators)) {
+      throw new Error([
+        'Invalid argument supplied to all, expected an instance of array.',
+        `See ${componentName} prop \`${propName}\`.`,
+      ].join(' '))
+    }
+
+    const errors = _.map(validators, validator => {
+      if (!_.isFunction(validator)) {
+        throw new Error(
+          `any() argument "validators" should contain functions, found: ${type(validator)}.`
+        )
+      }
+      return validator(props, propName, componentName, ...rest)
+    })
+
+    // if no validator passed return the first error found
+    if (!_.some(errors, _.overSome(_.isNull, _.isUndefined))) {
+      return _.find(errors, error => !_.isUndefined(error))
+    }
+  }
+}
+
+/**
+ * Ensure a validator passes only when a component has a given propsShape.
+ * @param {object} propsShape An object describing the prop shape.
+ * @param {function} validator A propType function.
+ */
+export const givenProps = (propsShape, validator) => {
+  return (props, propName, componentName, ...rest) => {
+    // only validate if all propsShape validators pass
+    if (!_.every(propsShape, (val, key) => !val(props, key, componentName, ...rest))) return
+
+    if (!_.isPlainObject(propsShape)) {
+      throw new Error(
+        `Invalid argument supplied to whenShape, expected an object. See ${componentName} prop \`${propName}\`.`,
+      )
+    }
+
+    if (!_.isFunction(validator)) {
+      throw new Error(
+        `Invalid argument supplied to whenShape, expected a function. See ${componentName} prop \`${propName}\`.`,
+      )
+    }
+
+    const error = validator(props, propName, componentName, ...rest)
+
+    if (error) {
+      // poor mans shallow pretty print, prevents JSON circular reference errors
+      const prettyProps = `{ ${_.map(_.pick(props, _.keys(propsShape)), (val, key) => {
+        let value = val
+        if (_.isString(val)) value = `"${val}"`
+        else if (_.isArray(val)) value = `[${val.join(', ')}]`
+        else if (_.isObject(val)) value = '{...}'
+
+        return `${key}: ${value}`
+      }).join(', ')} }`
+
+      error.message = `Given props ${prettyProps}: ${error.message}`
+      return error
+    }
+  }
+}
+
+/**
+ * Define prop dependencies by requiring other props.
+ * @param {string[]} requiredProps An array of required prop names.
+ */
+export const demand = (requiredProps) => {
```

Comment:
Renamed `require` => `demand` to avoid global collisions with node's require.  Borrowed from the  `yargs` library.